### DISCUSSION TITLE
Integrating external tools

### DISCUSSION THREAD

**Discussion by ipspace** [ipspace/netlab#770] [Features]
I always wanted to use _netlab_ to test external tools like SuzieQ. Graphite (currently implemented as an output module) would be in the same category.

In the ideal world, we'd have a framework that would generate tool configuration files from transformed lab topology, start the tool(s) at **netlab up**, and stop them at **netlab down**.

I documented a potential approach to that challenge in `docs/roadmap/tools.md` (also @ https://netsim-tools.readthedocs.io/en/dev/roadmap/tools.html). Comments would be highly appreciated (before I start coding the whole thing ;)

CC: @ddutt @ssasso @petercrocker @jbemmel

**Comment by ddutt** [comment]
Hi Ivan,

First of all, thanks for doing this. One concern with non-docker instances
would be the conflict in dependencies. If tool A wants version X of a
library but netlab wants version Y of thst dsmr lib, then this would be
very problematic. So, either simply support docker instances only OR
provide support for generating the appropriate configs and its upto the
user to use these separately, for example to startup the tool in a separate
virtualenv.

Thoughts?

Dinesh

**Reply by ipspace** [reply]
Good one, thank you!

* We could support virtual environments in tool configurations (it looks like a virtual environment is nothing more than setting PATH and VIRTUAL_ENV environment variables)
* We should have a `manual` runtime method that creates the configuration files but skips the tool startup/shutdown.

**Comment by ddutt** [comment]
Both sound like valid options to me, thank you. I presume that the env vars
will be set to run only within the tool startup script context and not
outside to ensure that erroneous runs don't result in a confused state for
the user where netlab commands start to fail due to missed up PATH and
VIRTUAL_ENV vars

Dinesh

**Reply by ipspace** [reply]
>  presume that the env vars will be set to run only within the tool startup script context

Absolutely.
